blockmanager.go

```diff
@@ -1019,109 +1022,153 @@ func (b *blockManager) getCheckpointedCFHeaders(checkpoints []*chainhash.Hash,
 			// they're too far ahead, or discard them if we don't
 			// need them.
 
-			// Find the first and last height for the blocks
-			// represented by this message.
-			startHeight := checkPointIndex*wire.CFCheckptInterval + 1
-			lastHeight := (nextCheckPointIndex + 1) * wire.CFCheckptInterval
+			select {
+			case headerChan <- r:
+			case <-b.quit:
+				return false, false
+			}
 
-			log.Debugf("Got cfheaders from height=%v to "+
-				"height=%v, prev_hash=%v", startHeight,
-				lastHeight, r.PrevFilterHeader)
+			return true, true
+		}
 
-			// If this is out of order but not yet written, we can
-			// verify that the checkpoints match, and then store
-			// them.
-			if startHeight > curHeight+1 {
-				log.Debugf("Got response for headers at "+
-					"height=%v, only at height=%v, stashing",
-					startHeight, curHeight)
+		msgs = append(msgs, &query.Query{
+			Req:        m,
+			HandleResp: handleResp,
+		})
+	}
 
-				queryResponses[checkPointIndex] = r
+	// Hand the queries to the work manager, and consume the verified
+	// responses as they come back.
+	errChan := b.cfg.QueryAccess.Query(msgs)
 
-				return true
-			}
+	// Keep waiting for more headers as long as we haven't received an
+	// answer for our last checkpoint, and no error is encountered.
+	for {
+		var r *wire.MsgCFHeaders
+		select {
+		case r = <-headerChan:
+		case err := <-errChan:
+			log.Errorf("Query finished with error (%v) before "+
```

Comment:
Nit: add error at the end to make it easier to parse?

Reply:
Also needs `if err != nil`.

Reply:
Yep, fixed: https://github.com/lightninglabs/neutrino/pull/179/commits/dab9f9f4046ea7711cbe1481ef4f52f7552f38f3